editoast/src/views/v2/timetable/stdcm.rs

```diff
@@ -233,6 +246,87 @@ async fn stdcm(
     Ok(Json(stdcm_response))
 }
 
+/// get the maximum run time, compute it if unspecified.
+/// returns an enum with either the result or a SimulationResponse if it failed
+async fn get_max_run_time(
+    db_pool: Arc<DbConnectionPool>,
+    redis_client: Arc<RedisClient>,
+    core_client: Arc<CoreClient>,
+    data: &STDCMRequestPayload,
+    infra: &Infra,
+    rolling_stock: &RollingStockModel,
+) -> Result<MaxRunningTimeResult> {
+    if data.maximum_run_time.is_some() {
+        return Ok(MaxRunningTimeResult::MaxRunningTime {
+            value: data.maximum_run_time.unwrap(),
+        });
+    }
+
+    let train_schedule = TrainSchedule {
+        id: 0,
+        train_name: "".to_string(),
+        labels: vec![],
+        rolling_stock_name: rolling_stock.name.clone(),
+        timetable_id: 0, // TODO: do we actually need a valid ID?
+        start_time: data.start_time,
+        schedule: vec![],
+        margins: build_single_margin(data.margin),
+        initial_speed: 0.0,
+        comfort: data.comfort,
+        path: convert_steps(&data.steps),
+        constraint_distribution: Default::default(),
+        speed_limit_tag: data.speed_limit_tags.clone(),
+        power_restrictions: vec![],
+        options: Default::default(),
+    };
+
+    // TODO: check that the result isn't actually saved in db
+    let sim_result =
+        train_simulation(db_pool, redis_client, core_client, &train_schedule, infra).await?;
+
+    let total_stop_time: u64 = data
+        .steps
+        .iter()
+        .map(|step: &PathfindingItem| step.duration.unwrap_or_default())
+        .sum();
+
+    return Ok(match sim_result {
+        SimulationResponse::Success { provisional, .. } => MaxRunningTimeResult::MaxRunningTime {
```

Comment:
How about adding this to avoid the `unwrap` in case where provisional is empty?
```suggestion
        SimulationResponse::Success { provisional, .. } if !provisional.is_empty() => MaxRunningTimeResult::MaxRunningTime {
```

Reply:
The unwrap isn't on `provisional` (can't be empty), it's on `provisional.times.last()`. As in, it would fail if the simulation doesn't contain any data at all.

There's really no situation where that *should* happen, but if it did anyway, returning an error from the simulation result wouldn't work out. In core I'd write an assert but I'm not sure that would result in a good response format, what would be the recommended way to handle that kind of assumption we're 99.9% sure are true?

Reply:
As discussed: the `unwrap` is now an `expect`. We don't really have a better way to handle really uncommon errors.

`core` now also has an assert for this case.

Reply:
As discussed, we might make a check on `core` to never return an empty list (and error in such case). But that wouldn't help on the current Rust code. So we might just convert the `unwrap` into `expect` so we can give a bit of context.